src/accountsdb/download.zig

```diff
@@ -538,10 +544,171 @@ pub fn downloadFile(
+        num_threads_snapshot_unpack: u16 = 0,
+        min_snapshot_download_speed_mbs: usize = 20,
+        trusted_validators: ?[]const Pubkey = null,
+    },
+) !struct { AllSnapshotFields, SnapshotFiles } {
+    const logger = logger_.withScope(LOG_SCOPE);
 
-const KeyPair = std.crypto.sign.Ed25519.KeyPair;
+    const force_unpack_snapshot = options.force_unpack_snapshot;
+    const force_new_snapshot_download = options.force_new_snapshot_download;
+    var n_threads_snapshot_unpack: u32 = options.num_threads_snapshot_unpack;
+    if (n_threads_snapshot_unpack == 0) {
+        const n_cpus = @as(u32, @truncate(try std.Thread.getCpuCount()));
+        n_threads_snapshot_unpack = n_cpus / 2;
+    }
+
+    // check if we need to download a fresh snapshot
+    var accounts_db_exists = blk: {
+        if (validator_dir.openDir("accounts_db", .{ .iterate = true })) |dir| {
+            std.posix.close(dir.fd);
+            break :blk true;
+        } else |_| {
+            break :blk false;
+        }
+    };
+
+    // clear old snapshots, if we will download a new one
+    if (force_new_snapshot_download and accounts_db_exists) {
+        logger.info().log("deleting accounts_db dir...");
+        try validator_dir.deleteTreeMinStackSize("accounts_db");
+        accounts_db_exists = false;
+    }
+
+    if (!accounts_db_exists) try validator_dir.makeDir("accounts_db");
+    const snapshot_dir = try validator_dir.openDir("accounts_db", .{});
+
+    // download a new snapshot if required
+    const snapshot_exists = !std.meta.isError(snapshot_dir.access("accounts", .{}));
+    const should_download_snapshot = force_new_snapshot_download or !snapshot_exists;
+    if (should_download_snapshot) {
+        const min_mb_per_sec = options.min_snapshot_download_speed_mbs;
+        const gossip_service = options.gossip_service orelse {
+            return error.SnapshotsNotFoundAndNoGossipService;
+        };
+
+        try downloadSnapshotsFromGossip(
+            allocator,
+            logger.unscoped(),
+            options.trusted_validators,
+            gossip_service,
+            snapshot_dir,
+            @intCast(min_mb_per_sec),
+        );
+    }
+
+    const valid_acounts_folder = blk: {
+        // NOTE: we only need to check this if we are *not* unpacking a fresh snapshot
+        if (force_unpack_snapshot or !snapshot_exists) break :blk false;
+
+        // do a quick sanity check on the number of files in accounts/
+        // NOTE: this is sometimes the case that you unpacked only a portion
+        // of the snapshot
+        var accounts_dir = snapshot_dir.openDir("accounts", .{}) catch |err| switch (err) {
+            // accounts folder doesnt exist, so its invalid
+            error.FileNotFound => break :blk false,
+            else => return err,
+        };
+        defer accounts_dir.close();
+        const n_account_files = (try accounts_dir.stat()).size;
+        if (n_account_files <= 100) {
```

Comment:
More of a question. Why is 100 the threshold to determine if the account dir is empty?

Reply:
its kinda a random threshold - if only 100 files exist then the snapshot was likely never fully unpacked (likely only partially unpacked and hit an error) but could also work for higher values (we usually expect 400k files for testnet/devnet/mainnet)

Reply:
Gotcha. It is indeed arbitrary, 101 files won't be considered partially unpacked, while it is not different from 100. But I also do not have any better idea for improving this...